TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/RoadSystemGraph.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using RoadGenerator;
+using UnityEngine;
+
+namespace RoadGenerator
+{
+    public class GraphEdge
+    {
+        public double Cost;
+        public GraphNode EndNode;
+        public GraphEdge(GraphNode endNode, double cost)
+        {
+            EndNode = endNode;
+            Cost = cost;
+        }
+    }
+    public class GraphNode
+    {
+        public RoadNode RoadNode;
+        public List<GraphEdge> Edges = new List<GraphEdge>();
+        public GraphNode(RoadNode roadNode)
+        {
+            this.RoadNode = roadNode;
+        }
+    }
+    /// <summary> A graph representation of a road </summary>
+    public class RoadNavigationGraph
+    {
+        public Dictionary<string, GraphNode> Graph = new Dictionary<string, GraphNode>();
+
+        // The nodes that should become part of the navigation graph
+        private readonly RoadNodeType[] _roadNodeTypesToAdd = 
+        {
+            RoadNodeType.ThreeWayIntersection,
+            RoadNodeType.FourWayIntersection,
+            RoadNodeType.End
+        };
+
+        private float _currentCost = 0f;
+        public RoadNavigationGraph(RoadNode roadNode)
+        {
+            RoadNode curr = roadNode;
+            GraphNode PreviouslyAddedNode = null;
+            while (curr != null)
+            {
+                // The start node is added to the graph
+                if (PreviouslyAddedNode == null)
+                {
+                    PreviouslyAddedNode = new GraphNode(curr);
+                    Graph.Add(curr.Position.ToString(), PreviouslyAddedNode);
+                    curr = curr.Next;
+                    continue;
+                }
+                _currentCost += CalculateCost(Vector3.Distance(curr.Position, curr.Prev.Position)); 
```

Comment:
This should be remade using the new `distanceToNextNode` being added to `RoadNode` in https://github.com/marc7s/TrafficSimulator/pull/35